File: objc/objcfind/Tests/objcfindTests/FileTypesTests.m
#import <XCTest/XCTest.h>
#import "FileTypes.h"

@interface FileTypesTests : XCTestCase
@property FileTypes *fileTypes;
@end

@implementation FileTypesTests

- (void)setUp {
    [super setUp];
    self.fileTypes = [[FileTypes alloc] init];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testArchiveFiles {
    NSArray<NSString*> *exts = @[@"7z", @"arj", @"bz2", @"cpio", @"ear", @"gz", @"hqx",
                                 @"jar", @"pax", @"rar", @"sit", @"sitx", @"tar", @"tgz", @"war",
                                 @"zip", @"zipx", @"Z"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"archive.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeArchive);
        XCTAssert([self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isImageFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testAudioFiles {
    NSArray<NSString*> *exts = @[@"aac", @"cda", @"mid", @"midi", @"mp3", @"oga", @"opus", @"wav", @"weba"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"audio.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeAudio);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert([self.fileTypes isAudioFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isImageFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testBinaryFiles {
    NSArray<NSString*> *exts = @[@"a", @"beam", @"chm", @"class", @"com", @"dat",
                                 @"dll", @"doc", @"dot", @"dylib", @"exe", @"hlp", @"lib", @"mdb",
                                 @"pdb", @"pot", @"so"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"binary.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeBinary);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isAudioFile:fileName]);
        XCTAssert([self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isImageFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testCodeFiles {
    NSArray<NSString*> *exts = @[@"asp", @"bas", @"bash", @"bat", @"c", @"clj", @"cpp", @"cs",
                                 @"css", @"erl", @"fs", @"go", @"groovy", @"h", @"hpp", @"hs",
                                 @"htm", @"html", @"java", @"js", @"m", @"php", @"py", @"rb",
                                 @"rc", @"scala", @"sh",  @"swift", @"ts", @"vb"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"code.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeCode);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isAudioFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert([self.fileTypes isCodeFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isImageFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testFontFiles {
    NSArray<NSString*> *exts = @[@"eot", @"otf", @"tte", @"ttf", @"woff", @"woff2"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"font.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeFont);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isAudioFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert([self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isImageFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testImageFiles {
//    NSArray<NSString*> *exts = @[@"ai", @"avif", @"bmp", @"bpg", @"cur", @"dib", @"drw"];
    NSArray<NSString*> *exts = @[@"avif", @"bmp", @"bpg", @"cur", @"dib", @"drw"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"image.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeImage);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isAudioFile:fileName]);
//        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert([self.fileTypes isImageFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testTextFiles {
    NSArray<NSString*> *exts = @[@"cfg", @"conf", @"csv", @"ddl", @"ini", @"log",
                                 @"markdown", @"md", @"po", @"properties", @"sgm", @"sgml",
                                 @"txt" , @"yml"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"text.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeText);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert([self.fileTypes isTextFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isVideoFile:fileName]);
    }
}

- (void)testVideoFiles {
    NSArray<NSString*> *exts = @[@"avi", @"mov", @"mp4", @"mpeg", @"ogv", @"webm"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"video.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeVideo);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isAudioFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert(![self.fileTypes isFontFile:fileName]);
        XCTAssert(![self.fileTypes isImageFile:fileName]);
        XCTAssert([self.fileTypes isVideoFile:fileName]);
    }
}

@end

File: swift/swiftfind/Tests/swiftfindTests/FileTypesTests.swift
//
//  FileTypesTests.swift
//  swiftfindTests
//
//  Created by Cary Clark on 5/18/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import XCTest

import swiftfind

class FileTypesTests: XCTestCase {
    let fileTypes = FileTypes()

//    override func setUp() {
//        super.setUp()
//    }
//
//    override func tearDown() {
//        super.tearDown()
//    }

    func testArchiveFiles() {
        let exts = ["7z", "arj", "bz2", "cpio", "ear", "gz", "hqx", "jar",
                    "pax", "rar", "sit", "sitx", "tar", "tgz", "war", "zip", "zipx",
                    "Z"]
        for ext in exts {
            let fileName = "archive.\(ext)"
            XCTAssert(fileTypes.isArchiveFile(fileName), "isArchiveFile(\(fileName)) == true")
            XCTAssert(!fileTypes.isBinaryFile(fileName), "isBinaryFile(\(fileName)) == false")
            XCTAssert(!fileTypes.isTextFile(fileName), "isTextFile(\(fileName)) == false")
            XCTAssert(fileTypes.getFileType(fileName) == FileType.archive,
                      "\(fileName) is archive file")
        }
    }

    func testAudioFile() {
        let fileName = "music.mp3"
        XCTAssert(fileTypes.isAudioFile(fileName), "isAudioFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.audio, "\(fileName) is audio file")
    }

    func testBinaryFiles() {
        let exts = ["a", "beam", "chm", "class", "com", "dat", "dbmdl", "dcr", "dir", "dll",
                    "dms", "doc", "dot", "dxr", "dylib", "epub", "exe", "fm", "hi", "hlp",
                    "lib", "lnk", "mdb", "mo", "mobi", "mpp", "nib", "o", "obj", "odm", "odt",
                    "ott", "pages", "pdb", "ppt", "pub", "pyc", "pyo", "qxd", "rpt", "so",
                    "swf", "sys", "vsd", "wpd", "wps", "wpt", "wri", "xls", "xlt"]
        for ext in exts {
            let fileName = "binfile.\(ext)"
            XCTAssert(!fileTypes.isArchiveFile(fileName), "isArchiveFile(\(fileName)) == false")
            XCTAssert(fileTypes.isBinaryFile(fileName), "isBinaryFile(\(fileName)) == true")
            XCTAssert(!fileTypes.isTextFile(fileName), "isTextFile(\(fileName)) == false")
            XCTAssert(fileTypes.getFileType(fileName) == FileType.binary,
                      "\(fileName) is binary file")
        }
    }

    func testCodeFile() {
        let fileName = "code.swift"
        XCTAssert(fileTypes.isCodeFile(fileName), "isCodeFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.code, "\(fileName) is code file")
    }

    func testFontFile() {
        let fileName = "font.ttf"
        XCTAssert(fileTypes.isFontFile(fileName), "isFontFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.font, "\(fileName) is font file")
    }

    func testImageFile() {
        let fileName = "image.png"
        XCTAssert(fileTypes.isImageFile(fileName), "isImageFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.image, "\(fileName) is image file")
    }

    func testTextFile() {
        let fileName = "text.txt"
        XCTAssert(fileTypes.isTextFile(fileName), "isTextFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.text, "\(fileName) is text file")
    }

    func testVideoFile() {
        let fileName = "movie.mp4"
        XCTAssert(fileTypes.isVideoFile(fileName), "isVideoFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.video, "\(fileName) is video file")
    }

    func testXmlFile() {
        let fileName = "markup.xml"
        XCTAssert(fileTypes.isXmlFile(fileName), "isXmlFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.xml, "\(fileName) is xml file")
    }

    func testUnknownFile() {
        let fileName = "unknown.ZZZ"
        XCTAssert(fileTypes.getFileType(fileName) == FileType.unknown, "\(fileName) is unknown file")
    }

    static let allTests = [
        ("testArchiveFiles", testArchiveFiles),
        ("testBinaryFiles", testBinaryFiles),
        ("testCodeFile", testCodeFile),
        ("testTextFile", testTextFile),
        ("testXmlFile", testXmlFile),
        ("testUnknownFile", testUnknownFile),
    ]
}
